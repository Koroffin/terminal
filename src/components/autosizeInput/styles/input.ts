import styled from 'styled-components';

export const Input = styled.input`
  font-family: monospace;
  width: 0;
`;
