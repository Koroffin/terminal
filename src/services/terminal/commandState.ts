export class CommandState {
  public status = 0;
  public output: string[] = [];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public parsedData: { [s: string]: any };
}
