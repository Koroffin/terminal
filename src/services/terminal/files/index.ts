import { generateCoolName } from 'Services/names';
import { DirectoryStruct, PrivateDirectoryStruct } from './directory';
import { FileStruct } from './file';

export const DEFAULT_DIR_COUNT = 3;
export const DEFAULT_FILE_COUNT = 5;
export const DEFAULT_SUBDIR_COUNT = 2;
export const DEFAULE_SUBFILES_COUNT = 3;

export class TerminalFiles {
  public static rootDir: PrivateDirectoryStruct = new PrivateDirectoryStruct();
  private static currentDir: DirectoryStruct = TerminalFiles.rootDir;
  public static get currentDirName(): string {
    return this.currentDir.name;
  }
  private static $makeDir(
    name: string = generateCoolName({ isDir: true }),
    files: FileStruct[] = []
  ): DirectoryStruct {
    return new DirectoryStruct(name, files);
  }
  private static $makeFile(name: string = generateCoolName()): FileStruct {
    return new FileStruct(name);
  }
  public static generateDefaultTree() {
    for (let i = 0; i < DEFAULT_DIR_COUNT; i++) {
      const dir = TerminalFiles.$makeDir();

      for (let j = 0; j < DEFAULT_SUBDIR_COUNT; j++) {
        const subDir = TerminalFiles.$makeDir();
        for (let k = 0; k < DEFAULE_SUBFILES_COUNT; k++) {
          const file = TerminalFiles.$makeFile();
          subDir.addFile(file);
        }
        dir.addDir(subDir);
      }

      for (let j = 0; j < DEFAULT_FILE_COUNT; j++) {
        const file = TerminalFiles.$makeFile();
        dir.addFile(file);
      }

      TerminalFiles.rootDir.addDir(dir);
    }
  }
  public static ls(sortBy: keyof FileStruct = 'name'): string[] {
    const res: string[] = ['.'];
    if (this.currentDir !== TerminalFiles.rootDir) {
      res.push('..');
    }
    const dir = TerminalFiles.currentDir;
    dir.forEachDir(({ name }) => res.push(`${name}/`), sortBy);
    dir.forEachFile(({ name }) => res.push(name), sortBy);
    return res;
  }
  public static cd(dir: string) {
    const dirStruct = TerminalFiles.currentDir.getDir(dir);
    if (dirStruct) {
      TerminalFiles.currentDir = dirStruct;
    }
  }
  public static reset() {
    TerminalFiles.rootDir = new PrivateDirectoryStruct();
    TerminalFiles.currentDir = TerminalFiles.rootDir;
  }
}
