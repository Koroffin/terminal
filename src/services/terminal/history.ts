export class TerminalHistory {
  public static history: string[] = [];
  private static pointer = 0;

  public static add(s: string) {
    TerminalHistory.history.push(s);
    TerminalHistory.pointer = TerminalHistory.l;
  }
  private static getByPointer(): string {
    return TerminalHistory.history[TerminalHistory.pointer] ?? '';
  }
  public static prev(): string {
    TerminalHistory.pointer = Math.max(0, TerminalHistory.pointer - 1);
    return TerminalHistory.getByPointer();
  }
  public static next(): string {
    TerminalHistory.pointer = Math.min(
      TerminalHistory.l,
      TerminalHistory.pointer + 1
    );
    return TerminalHistory.getByPointer();
  }
  public static reset() {
    TerminalHistory.history = [];
    TerminalHistory.pointer = 0;
  }
  public static get l(): number {
    return TerminalHistory.history.length;
  }
  public static removeByIdx(idx: number) {
    TerminalHistory.history.splice(idx, 1);
  }
}
