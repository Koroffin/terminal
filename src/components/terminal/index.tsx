import React, { useCallback, useEffect, useState } from 'react';
import { TerminalService } from 'Services/terminal';
import { Input } from './input';
import { Wrapper } from './styles/wrapper';
import { nanoid } from 'nanoid';
import { Log } from './interfaces/log';
import { Block } from './styles/block';

export const Terminal: React.FC = () => {
  const [log, setLog] = useState<Log[]>([]);
  useEffect(() => TerminalService.init(), []);

  const execute = useCallback(
    (str: string) => {
      const output = TerminalService.parse(str);
      setLog([
        ...log,
        { id: nanoid(), value: `$ ${str}` },
        ...output.map((value) => ({ id: nanoid(), value })),
      ]);
    },
    [setLog, log]
  );
  const onClick = useCallback((evt: React.SyntheticEvent<HTMLDivElement>) => {
    const { target } = evt;
    if (target instanceof HTMLDivElement) {
      const input = target.getElementsByTagName('input')[0];
      input.focus();
    }
  }, []);

  return (
    <Wrapper onClick={onClick}>
      {log.map((item) => (
        <Block key={item.id}>{item.value}</Block>
      ))}
      <Input execute={execute} />
    </Wrapper>
  );
};
