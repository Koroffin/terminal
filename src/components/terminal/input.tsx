import React, { useCallback, useState } from 'react';
import { TerminalService } from 'Services/terminal';
import { Flex } from 'Components/flex';
import { Block } from './styles/block';
import { Caret } from './styles/caret';
import { InputWrapper } from './styles/inputWrapper';

interface InputProps {
  execute(str: string): void;
}

export const Input: React.FC<InputProps> = ({ execute }) => {
  const [isFocused, setIsFocused] = useState<boolean>(false);
  const [caretPosition, setCaretPosition] = useState<number>(0);

  const updateCaretPosition = useCallback(
    (input: HTMLInputElement) => {
      setCaretPosition(input.selectionEnd);
    },
    [setCaretPosition]
  );
  const updateInputContent = useCallback(
    (target: HTMLInputElement, newValue: string) => {
      const caretPos = newValue.length;
      target.value = newValue;
      target.setSelectionRange(caretPos, caretPos);
    },
    []
  );

  const onFocus = useCallback(
    (evt: React.SyntheticEvent<HTMLInputElement>) => {
      const { target } = evt;
      if (target instanceof HTMLInputElement) {
        updateCaretPosition(target);
      }
      setIsFocused(true);
    },
    [setIsFocused, updateCaretPosition]
  );
  const onBlur = useCallback(() => setIsFocused(false), [setIsFocused]);

  const onInputClick = useCallback(
    (evt: React.SyntheticEvent<HTMLInputElement>) => {
      const { target } = evt;
      if (target instanceof HTMLInputElement) {
        updateCaretPosition(target);
      }
    },
    [updateCaretPosition]
  );

  const onKeyDown = useCallback(
    (evt: React.KeyboardEvent<HTMLInputElement>) => {
      const { target, key } = evt;
      if (target instanceof HTMLInputElement) {
        if (key === 'Enter') {
          execute(target.value);
          target.value = '';
        }
        if (key === 'ArrowUp') {
          requestAnimationFrame(() =>
            updateInputContent(target, TerminalService.history.prev())
          );
        }
        if (key === 'ArrowDown') {
          requestAnimationFrame(() =>
            updateInputContent(target, TerminalService.history.next())
          );
        }
        requestAnimationFrame(() => updateCaretPosition(target));
      }
    },
    [updateCaretPosition, execute, updateInputContent]
  );

  return (
    <Block>
      <Flex>
        $&nbsp;
        <InputWrapper
          key="input"
          onFocus={onFocus}
          onBlur={onBlur}
          onClick={onInputClick}
          onKeyDown={onKeyDown}
          autoFocus
        />
        {isFocused && <Caret position={caretPosition} />}
      </Flex>
    </Block>
  );
};
