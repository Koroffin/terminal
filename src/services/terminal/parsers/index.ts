import { CommandState } from 'Services/terminal/commandState';
import { regexp } from './utils/regexp';

export class AbstractParser {
  public static command = 'abstract';
  public static canBeSingleWord = true;
  public static regexp(): RegExp {
    return regexp(this.command, { canBeSingleWord: this.canBeSingleWord });
  }
  public static test(s: string): boolean {
    return this.regexp().test(s);
  }
  public static parse(_s: string): CommandState['parsedData'] {
    return {};
  }
  public static execute(_state: CommandState) {}
  public static parseFlags(str: string): string[] {
    return (str.match(/\s--?([A-Za-z0-9-]+)/g) ?? [])
      .map((f) => f.trim())
      .reduce((acc, f) => {
        if (f.startsWith('--')) {
          acc.push(f);
          return acc;
        }

        return acc.concat(
          f
            .substring(1)
            .split('')
            .map((c) => `-${c}`)
        );
      }, []);
  }
}
