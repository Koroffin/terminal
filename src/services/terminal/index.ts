import { CommandState } from './commandState';
import { TerminalFiles } from './files';
import { TerminalHistory } from './history';

import { GrepParser } from './parsers/grep';
import { HistoryParser } from './parsers/history';
import { LsParser } from './parsers/ls';
import { ManParser } from './parsers/man';
import { SudoParser } from './parsers/sudo';

export class TerminalService {
  public static history = TerminalHistory;
  public static files = TerminalFiles;
  private static parsers = [
    SudoParser,
    HistoryParser,
    ManParser,
    GrepParser,
    LsParser,
  ];
  private static sanitize(str: string): string {
    return str.replace(/\s+/g, ' ').trim();
  }
  public static init() {
    TerminalService.files.generateDefaultTree();
  }
  public static reset() {
    TerminalService.files.reset();
    TerminalService.history.reset();
  }
  public static parse(str: string): string[] {
    TerminalService.history.add(str);
    const { parsers } = TerminalService;
    const cleanStr = TerminalService.sanitize(str);
    const commands = cleanStr.split('|').map((s) => s.trim());
    const state = new CommandState();
    for (let i = 0, l = commands.length; i < l; i++) {
      const command = commands[i];
      for (let j = 0, _l = parsers.length; j < _l; j++) {
        const parser = parsers[j];
        if (parser.test(command)) {
          state.parsedData = parser.parse(command);
          parser.execute(state);
        }
      }
    }
    return state.output;
  }
  public static autocomplete(str: string): string[] {
    const { parsers } = TerminalService;
    const cleanStr = TerminalService.sanitize(str);
    const res: string[] = [];
    for (let i = 0, l = parsers.length; i < l; i++) {
      const { command } = parsers[i];
      if (command.startsWith(cleanStr)) {
        res.push(command);
      }
    }
    return res.sort((a, b) => a.localeCompare(b));
  }
}
