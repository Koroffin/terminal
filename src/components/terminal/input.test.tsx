//write tests
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Input } from './input';
import { TerminalService } from 'Services/terminal';
import { act } from 'react-dom/test-utils';

describe('Input', () => {
  it('should render correctly', () => {
    const { container } = render(<Input execute={jest.fn()} />);
    expect(container).toMatchSnapshot();
  });
  it('should call execute when Enter key is pressed', () => {
    const execute = jest.fn();
    const { getByTestId } = render(<Input execute={execute} />);
    const input = getByTestId('autosize-input');
    fireEvent.keyDown(input, { key: 'Enter' });
    expect(execute).toHaveBeenCalled();
  });
  it('should call TerminalService.history.prev when ArrowUp key is pressed', () => {
    const spyOnPrev = jest.spyOn(TerminalService.history, 'prev');
    const { getByTestId } = render(<Input execute={jest.fn()} />);
    const input = getByTestId('autosize-input');
    fireEvent.keyDown(input, { key: 'ArrowUp' });
    expect(spyOnPrev).toHaveBeenCalled();
  });
  it('should call TerminalService.history.next when ArrowDown key is pressed', () => {
    const spyOnNext = jest.spyOn(TerminalService.history, 'next');
    const { getByTestId } = render(<Input execute={jest.fn()} />);
    const input = getByTestId('autosize-input');
    fireEvent.keyDown(input, { key: 'ArrowDown' });
    expect(spyOnNext).toHaveBeenCalled();
  });
  it('should render caret on focus', () => {
    const { getByTestId, queryByTestId } = render(
      <Input execute={jest.fn()} />
    );
    const input = getByTestId('autosize-input');
    act(() => {
      fireEvent.focus(input);
    });
    expect(queryByTestId('caret')).toBeDOMElement();
  });
  it('should hide caret on blur', () => {
    const { getByTestId, queryByTestId } = render(
      <Input execute={jest.fn()} />
    );
    const input = getByTestId('autosize-input');
    act(() => {
      fireEvent.focus(input);
    });
    act(() => {
      fireEvent.blur(input);
    });
    expect(queryByTestId('caret')).toBeNull();
  });
  it('input is focused by default', () => {
    const { getByTestId } = render(<Input execute={jest.fn()} />);
    const input = getByTestId('autosize-input');
    expect(input).toBeFocused();
  });
});
