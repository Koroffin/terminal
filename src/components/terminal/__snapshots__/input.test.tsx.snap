// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly 1`] = `
<div>
  <div
    class="sc-gsnTZi tugAn"
  >
    <div
      class="sc-bczRLJ emMGHK"
    >
      $Â 
      <input
        class="sc-hKMtZM bMwtCL sc-eCYdqJ jdWWWf"
        data-autotest-id="autosize-input"
      />
      <div
        class="sc-dkzDqf cUEhTp"
        data-autotest-id="caret"
      />
    </div>
  </div>
</div>
`;
